integration/prefect/openlineage/prefect/adapter.py

```diff
@@ -0,0 +1,229 @@
+        """
+        event = RunEvent(
+            eventType=RunState.COMPLETE,
+            eventTime=end_time,
+            run=self._build_run(run_id),
+            job=self._build_job(job_name),
+            inputs=inputs,
+            outputs=outputs,
+            producer=_PRODUCER,
+        )
+        self.client.emit(event)
+
+    def fail_task(
+        self,
+        run_id: str,
+        job_name: str,
+        end_time: str,
+        inputs: Optional[Any],
+        outputs: Optional[Any],
+    ):
+        """
+        Emits openlineage event of type FAIL
+        :param run_id: globally unique identifier of task in dag run
+        :param job_name: globally unique identifier of task between dags
+        :param end_time: time of task completion
+        :param inputs: List of Input Datasets
+        :param outputs: List of Output Datasets
+        """
+        event = RunEvent(
+            eventType=RunState.FAIL,
+            eventTime=end_time,
+            run=self._build_run(run_id),
+            job=self._build_job(job_name),
+            inputs=inputs,
+            outputs=outputs,
+            producer=_PRODUCER,
+        )
+        self.client.emit(event)
+
+    @staticmethod
+    def _to_json(data: Dict):
+        cleaned = Serde.remove_nulls_and_enums(data)
+        return json.dumps(cleaned, sort_keys=True)
+
+    def list_sources(self):
+        url = urljoin(self.client.url, f"api/v1/sources/")
+        r = self.client.session.get(
+            url=url,
+            timeout=self.client.options.timeout,
+            verify=self.client.options.verify,
+        )
+        return r.json()['sources']
+
+    def create_source(
+        self,
+        source_name: str,
+        source_type: str,
+        connection_url: str,
+        description: Optional[str] = None,
+    ):
+        url = urljoin(self.client.url, f"api/v1/sources/{source_name}")
+        data = {"type": source_type, "connectionUrl": connection_url, "description": description}
+        r = self.client.session.put(
+            url=url,
+            data=self._to_json(data),
+            timeout=self.client.options.timeout,
+            verify=self.client.options.verify,
+        )
+        r.raise_for_status()
+        return r.status_code == 200
+
+    def create_dataset(
+        self,
+        source_name: str,
+        dataset_name: str,
+        url_path: str,
+        dataset_type: Optional[str] = "DB_TABLE",
+        namespace: Optional[str] = None,
+        **kwargs,
+    ):
+        namespace = namespace or self.namespace
+        # TODO - type might be dropped?
+        data = {
+            "type": dataset_type,
+            "physicalName": url_path,
+            "sourceName": source_name,
+            "fields": kwargs.pop("fields", []),
+            **kwargs,
+        }
+        r = self.client.session.put(
+            urljoin(self.client.url, f"api/v1/namespaces/{namespace}/datasets/{dataset_name}"),
```

Comment:
you don't need to create the datasets. You should just send OpenLineage RunEvents.
Same for sources.
See:
https://openlineage.io/getting-started#step-1-start-a-run

Additional information (schema) can be added in facets.

Reply:
I'm using this in the my  "ExternalData" task: 

```
class ExternalDataset(Task):
    """
    Task for reading data via `fsspec` and creating an OpenLineage Dataset at the same time.

    Useful if you want a dataset to be available in OpenLineage for lineage purposes, but don't have control over the
    process writing the data, and therefore don't have lineage already set up. An example would be some files on a HTTP
    server or S3 bucket.
```

What would you recommend in this situation @julienledem ? Is there some other equivalent in openlineage at this stage?

Reply:
There is not, so this is Marquez specific. It would be worth starting a [proposal] github issue on OpenLineage to discuss this.

Reply:
https://github.com/OpenLineage/OpenLineage/issues/294

Reply:
Sorry for the back and forth on that issue. Yes the dataset will be created on read so that works.